lib/core/src/Ouroboros/Network/Client/Wallet.hs

```diff
@@ -340,24 +355,61 @@ chainSyncWithBlocks (Quantity epochStability) fromTip queue =
             let tip'    = fromTip tip
             respond (RollForward cursor' tip' blocks')
             let distance = tipDistance (blockNo block) tip
-            let strategy = if distance > k + maxInFlight
-                    then pipeline Zero
-                    else oneByOne
+            let strategy = if distance <= 1
+                    then oneByOne
+                    else pipeline (fromIntegral $ min distance 1000) Zero
             clientStIdle strategy
 
-        , P.recvMsgRollBackward = \point _tip -> do
-            respond (RollBackward point)
-            clientStIdle oneByOne
+        -- When the last message we receive is a request to rollback, we have
+        -- two possibilities:
+        --
+        -- a) Either, we are asked to rollback to a point that is within the
+        -- blocks we have just collected. So it suffices to remove blocks from
+        -- the list, and apply the remaining portion.
+        --
+        -- b) We are asked to rollback even further and discard all the blocks
+        -- we just collected. In which case, we simply discard all blocks and
+        -- rollback to that point as if nothing happened.
+        , P.recvMsgRollBackward = \point tip ->
+            case rollback point blocks of
+                [] -> do -- b)
+                    respond (RollBackward point)
+                    clientStIdle oneByOne
+
+                xs -> do -- a)
+                    let cursor' = blockPoint $ head xs
+                    let blocks' = reverse xs
+                    let tip'    = fromTip tip
+                    respond (RollForward cursor' tip' blocks')
+                    clientStIdle oneByOne
         }
+
     collectResponses respond blocks (Succ n) = P.ClientStNext
         { P.recvMsgRollForward = \block _tip -> pure $
             P.CollectResponse Nothing $ collectResponses respond (block:blocks) n
-        , P.recvMsgRollBackward = \_point _tip -> pure $ error
-            "rolled backward while pipelining requests! This shouldn't be \
-            \possible because we are making sure to only pipeline requests \
-            \in areas where the node can't possibly roll back."
+
+        -- This scenario is slightly more complicated than for the 'Zero' case.
+        -- Again, there are two possibilities:
+        --
+        -- a) Either we rollback to a point we have just collected, so it
+        -- suffices to discard blocks from the list and continue.
+        --
+        -- b) Or, we need to reply immediately, but we still have to collect the
+        -- remaining responses. BUT, we can only reply once to a given command.
+        -- So instead, we stash all the remaining responses in a queue and, upon
+        -- receiving future requests, we'll simply read them from the stash!
+        , P.recvMsgRollBackward = \point _tip ->
+            case rollback point blocks of
+                [] -> do -- b)
```

Comment:
Yes, this would induce a rollback to where the wallet already is at, which is a no-op for the wallet and database. So, in principle, there's nothing wrong from it. Yet indeed, the client is blind to what the latest persisted block is so it can't really make this distinction.